### PR TITLE
enhance: Optimize DescribeIndex to reduce lock contention

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #29313 
issue: #30443

### CODE REVIEW

**Comment by czs007** [comment]
please add performance test to verify this modification.

**Comment by xiaocai2333** [comment]
Under the same conditions, describe index performance comparison：
- 10263 segments

for master branch, describe index performance: about 70+ms
```
[2024/03/01 10:41:02.427 +08:00] [INFO] [datacoord/index_service.go:603] ["receive DescribeIndex request"] [traceID=9f00c083d5284a5f42c0af90285d5008] [collectionID=448072142336557220] [indexName=] [timestamp=0]
[2024/03/01 10:41:02.500 +08:00] [INFO] [datacoord/index_service.go:650] ["DescribeIndex success"] [traceID=9f00c083d5284a5f42c0af90285d5008] [collectionID=448072142336557220] [indexName=]

describe index time cost: 0.07441520690917969 seconds
```

after optimized: about 10ms
```
[2024/03/01 10:23:47.801 +08:00] [INFO] [datacoord/index_service.go:657] ["receive DescribeIndex request"] [traceID=4484c6cfad36c5ffbfc96491e0de42c5] [collectionID=448072142336557220] [indexName=] [timestamp=0]
[2024/03/01 10:23:47.810 +08:00] [INFO] [datacoord/index_service.go:720] ["DescribeIndex success"] [traceID=4484c6cfad36c5ffbfc96491e0de42c5] [collectionID=448072142336557220] [indexName=]

describe index cost:  0.01020503044128418
```

File: internal/datacoord/index_service.go [hunk 1]
```diff
@@ -594,6 +607,15 @@ func (s *Server) GetIndexBuildProgress(ctx context.Context, req *indexpb.GetInde
 	}, nil
 }
 
+type indexInfoMeta struct {
+	ID             int64
+	numRows        int64
+	compactionFrom []int64
+	segmentIndexes map[int64]*model.SegmentIndex
```
Comment (czs007) [inline internal/datacoord/index_service.go:614]: simplify this map.

File: internal/datacoord/index_service.go [hunk 2]
```diff
@@ -594,6 +607,15 @@ func (s *Server) GetIndexBuildProgress(ctx context.Context, req *indexpb.GetInde
 	}, nil
 }
 
+type indexInfoMeta struct {
```
Comment (czs007) [inline internal/datacoord/index_service.go:610]: refine this name?

File: internal/datacoord/meta.go [hunk 3]
```diff
@@ -954,6 +955,33 @@ func (m *meta) SelectSegments(selector SegmentInfoSelector) []*SegmentInfo {
 	return ret
 }
 
+func (m *meta) SelectSegmentIndexes(selector SegmentInfoSelector) map[int64]*indexStats {
```
Comment (czs007) [inline internal/datacoord/meta.go:958]: SelectSegmentIndexesStats?

File: internal/datacoord/index_service.go [hunk 4]
```diff
@@ -478,42 +484,45 @@ func (s *Server) completeIndexInfo(indexInfo *indexpb.IndexInfo, index *model.In
 		pendingIndexRows = int64(0)
 	)
 
-	for _, seg := range segments {
-		totalRows += seg.NumOfRows
-		segIdx, ok := seg.segmentIndexes[index.IndexID]
+	for segID, seg := range segments {
+		if seg.state != commonpb.SegmentState_Flushed && seg.state != commonpb.SegmentState_Flushing {
```
Comment (czs007) [inline internal/datacoord/index_service.go:488]: !isFlush()?

**Pull request merged**